docroot/modules/humsci/hs_field_helpers/hs_field_helpers.module

```diff
@@ -725,3 +727,116 @@ function hs_field_helpers_field_widget_complete_link_default_form_alter(&$field_
   }
 
 }
+
+/**
+ * Implements hook_entity_form_display_alter().
+ */
+function hs_field_helpers_entity_form_display_alter(EntityFormDisplayInterface $form_display, array $context) {
+  // In order to disable the fields rather than use readonly_field_widget
+  // we need to remove the readonly_field_widget type and disable the fields.
+  $node = \Drupal::routeMatch()->getParameter('node');
+  if ($context['entity_type'] != 'node' || !$node) {
+    return;
+  }
+
+  $imported_fields = _hs_field_helpers_get_imported_fields($node);
+
+  if (!empty($imported_fields)) {
+    \Drupal::messenger()
+      ->addWarning(t('Some fields can not be edited since they contain imported & synced data.'));
+  }
+
+  foreach ($imported_fields as $field_name) {
+    // Remove the readonly_field_widget type.
+    $component['type'] = NULL;
+    $form_display->setComponent($field_name, $component);
+  }
+}
+
+/**
+ * Implements hook_entity_form_FORM_ID_alter().
+ */
+function hs_field_helpers_form_node_form_alter(&$form, FormStateInterface &$form_state, $form_id) {
+
+  // Disables the field if it's imported.
+  $node = \Drupal::routeMatch()->getParameter('node');
+  if (!$node || !$node instanceof NodeInterface) {
+    return;
+  }
+
+  $imported_fields = _hs_field_helpers_get_imported_fields($node);
+
+  if (!empty($imported_fields)) {
+    $form['#attributes']['data-has-imported-fields'][] = 'true';
+  }
+
+  foreach ($imported_fields as $field_name) {
+    if (isset($form[$field_name])) {
+      // Set the field to disabled.
+      $form[$field_name]['#disabled'] = TRUE;
+    }
+  }
+}
+
+/**
+ * Get the fields that are imported from the migration.
+ *
+ * @param \Drupal\node\NodeInterface $node
+ *   The node to get the imported fields from.
+ *
+ * @return array
+ *   The fields that are imported from the migration.
+ */
+function _hs_field_helpers_get_imported_fields(NodeInterface $node) : array {
+
```

Comment:
```suggestion
// Static cache for imported fields by node ID.
  static $imported_fields_cache = [];
  $nid = $node->id();
  if (isset($imported_fields_cache[$nid])) {
    return $imported_fields_cache[$nid];
  }
```
Since this is getting called at least twice in two separate hooks, a static cache of the fields here couldn't hurt.